proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,305 @@
+
+(There's a possiblity of *one-to-many* which a single downstream routes to several upstreams as if they were one, which might be useful for presenting several
+kafka clusters as if they were one, however, supporting transactions across two clusters would be difficult.  We consider won't this use-case futrther).
+
+## Current situation
+
+Currently Kroxyilicious binds a single port to the downstream and connects to a single upstream broker.  There is no support for clusters comprised of more
+then one broker.
+
+## Motivation
+
+1. Ability for Kroxylicious to expose multiple physical kafka clusters.
+1. Ability for Kroxylicious to support clusters formed of more than one broker.
+1. Ability for Kroxylicious to provide a building block to support multi-tenancy.
+
+## Proposal
+
+The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.
+
+### Key Objects
+
+
+![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)
+
+
+
+
+
+#### Endpoints
+
+Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  
+
+* For the non-TLS case, multiple endpoints are required to support >1 brokers. 
+* For TLS case, SNI gives a mechanism to route the traffic to a >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.
+
+`ProxyConfig` will be replaced by an list of _Endpoint_ objects.  
+
+```
+- name: endpoint name
+  address: bind address/port
+  type: PLAIN|TLS
+```
+
+#### Downstream Clusters
+
+Conceptually the _downstream cluster_ represents the cluster that the kafka client connects to.  The downstream clusters list specifies
+all the kafka clustsers are being presented by Kroxylicious. 
+
+A _downstream clustser_ references exactly one _upstream cluster_. 
+
+A _downstream cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the upstream cluster.
+
+A _downstream cluster_ enumerates the downstream brokers that comprise the downstream cluster.
+
+##### Downstream Brokers
+
+Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
+for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.
+
+Each downstream broker references exactly one upstream broker which must exist within the referenced upstream cluster.  It is an error if there is not a
+one-to-one correspondence between downstream and upstream brokers.
+
+TLS key material may be provided either at the downstream cluster level or at the individual broker level.
```

Comment:
What would the cluster-level TLS configuration be used for? We know the broker SNI hostnames, but would we know the bootstrap hostname to apply it to? Should a public hostname be part of the downstream cluster configuration?

Reply:
It is for the certificate per cluster use-case. The cert's SAN would need to list the FQDN of each broker within the cluster.

Reply:
Okay, I was more thinking of if we have a separate bootstrap host like:
`my-kafka-a.b.c.robeyoun-kafka.public.host`
and then per-broker hosts like `broker-1.my-kafka-a.b.c.robeyoun-kafka.public.host`

Where would we know how to map the bootstrap host to the downstream cluster SSL configuration?

Reply:
edit: maybe we can model bootstrap as a downstream and upstream broker.